lib/segment/client/http.ex

```diff
@@ -0,0 +1,150 @@
+defmodule Segment.Analytics.AdapterStub do
+  require Logger
+
+  def call(env, _opts) do
+    Logger.debug("[Segment] HTTP API called with #{inspect(env)}")
+    {:ok, %{env | status: 200, body: ""}}
+  end
+end
+
+defmodule Segment.Http do
+  require Logger
+  use Retry
+
+  @type client() :: Tesla.Client.t()
+
+  @segment_api_url "https://api.segment.io/v1/"
+  @send_to_http Application.get_env(:segment, :send_to_http, true)
+  @retry_attempts Application.get_env(:segment, :retry_attempts, 3)
+  @retry_expiry Application.get_env(:segment, :retry_expiry, 10_000)
+  @retry_start Application.get_env(:segment, :retry_start, 100)
+
+  @spec client(String.t()) :: client()
+  def client(api_key) do
+    adapter =
+      case @send_to_http do
+        true ->
+          Application.get_env(:segment, :tesla)[:adapter] ||
+            {Tesla.Adapter.Hackney, [recv_timeout: 30_000]}
+
+        false ->
+          {Segment.Analytics.AdapterStub, []}
+      end
+
+    client(api_key, adapter)
+  end
+
+  @spec client(String.t(), any()) :: client()
+  def client(api_key, adapter) do
+    middleware = [
+      {Tesla.Middleware.BaseUrl, @segment_api_url},
+      Tesla.Middleware.JSON,
+      {Tesla.Middleware.BasicAuth, %{username: api_key, password: ""}}
+    ]
+
+    Tesla.client(middleware, adapter)
+  end
+
+  def call(client, events) when is_list(events), do: batch(client, events)
+
+  def call(client, event) do
+    case callp(client, event.type, prepare_events(event), @retry_attempts) do
+      {:ok, %{status: status}} when status == 200 ->
+        :ok
+
+      {:ok, %{status: status}} when status == 400 ->
+        Logger.error("[Segment] Call Failed. JSON too large or invalid")
+        :error
+
+      {:error, err} ->
+        Logger.error("[Segment] Call Failed after #{@retry_attempts} retries. #{inspect(err)}")
+        :error
+
+      err ->
+        Logger.error("[Segment] Call Failed #{inspect(err)}")
+        :error
+    end
+  end
+
+  def batch(client, events, context \\ nil, integrations \\ nil) do
+    data =
+      %{batch: prepare_events(events)}
+      |> add_if(:context, context)
+      |> add_if(:integrations, integrations)
+
+    case callp(client, "batch", data, @retry_attempts) do
```

Comment:
Can likely consolidate these errors with the single non batch version -- also wondering thinking this could be a nice place to hook in a callback module that could handle the failed batch such that the user could store it for later or re-enqueue

Reply:
hmm. not sure i understand. might need you to write an example of what you are thinking

Reply:
Leave this for 0.3.0 ;)